nixos/modules/virtualisation/digital-ocean-init.nix

```diff
@@ -0,0 +1,63 @@
+{ config, pkgs, lib, ... }:
+with lib;
+{
+  options.virtualisation.digitalOcean.rebuildFromUserData = mkOption {
+    type = bool;
+    enable = true;
+    example = true;
+    description = "Whether to reconfigure the system from Digital Ocean user data";
+  };
+
+  config = {
+    systemd.services.digital-ocean-init = {
+      description = "Reconfigure the system from Digital Ocean uesrdata on startup";
+      wantedBy = [ "multi-user.target" ];
+      unitConfig = {
+        ConditionFileExists = "!/etc/nixos/configuration.nix";
+        After = [ "network-online.target" ];
+        Wants = [ "network-online.target" ];
+      };
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+      };
+      restartIfChanged = false;
+      path = [ pkgs.curl pkgs.gnused pkgs.gnugrep pkgs.systemd config.nix.package config.system.build.nixos-rebuild ];
+      script = ''
+        set -e
+        echo "attempting to fetch configuration from Digital Ocean user data..."
+        export HOME=/root
+        export NIX_PATH=/nix/var/nix/profiles/per-user/root/channels/nixos:nixos-config=/etc/nixos/configuration.nix:/nix/var/nix/profiles/per-user/root/channels
+        userData=$(mktemp)
+        if curl --retry-connrefused --output $userData http://169.254.169.254/metadata/v1/user-data; then
+          # If the user-data looks like it could be a nix expression,
+          # copy it over. Also, look for a magic three-hash comment and set
+          # that as the channel.
+          if sed '/^\(#\|SSH_HOST_.*\)/d' < "$userData" | grep -q '\S'; then
+            channels="$(grep '^###' "$userData" | sed 's|###\s*||')"
+            printf "%s" "$channels" | while read channel; do
+              echo "writing channel: $channel"
+            done
+
+            if [[ -n "$channels" ]]; then
+              printf "%s" "$channels" > /root/.nix-channels
+              nix-channel --update
+            fi
+
+            echo "setting configuration from Digital Ocean user data"
+            cp "$userData" /etc/nixos/configuration.nix
```

Comment:
I'm not entirely happy about this part. I'd like to find a way to make sure that digital-ocean-config is in the module list without having to explicitly include it. That way if a user provides a config in the user data they can assume still that the digital ocean stuff is in place, and disable it via the options if necessary.

cc @arianvp thoughts?

Reply:
Usually the pattern that people are used to , is:

```
# configuration.nix
{ imports = [ ./hardware-configuration.nix ]
}
```

right? At least, that is the standard `/etc/nixos/configuration.nix` that is being generated.

then just put the DO specific things in `/etc/nixos/hardware-configuration.nix`

It's not ideal because if people forget, the config won't work... but it is the same as on their laptops.


Other thing we could do is:


```
cp $userData /etc/nixos/do-userdata.nix
```

And then hardcode `/etc/nixos/configuration.nix` to be:

```
{ imports = [./do-userdata.nix  <nixpkgs....blah../do-config.nix> ]; }
```

Then people don't have to remember to include the `hardware-configuration` manually

Reply:
I was thinking along those lines too. But maybe we should stick to what the AWS/GCE images do, which is to put the user data in `/etc/nixos/configuration.nix`, and then make an issue to discuss this. Or we could just lead the way in doing things in a bit nicer way.

Reply:
Actually it would be nice to get @Infinisil's take as the code owner on this question.

Reply:
What @arianvp suggested sounds much better, I don't think there's much to discuss there. I'd just go ahead and do it this way, which can serve as an example for others.

Reply:
@eamsden shall we go with my suggestion? I'd love to see this merged. Been using it for quite a while already on my own nixpkgs fork

Reply:
@arianvp Yeah go with your suggestion. I too really want to see this merged